src/scheduler/controller.rs

```diff
@@ -131,7 +131,16 @@ impl<VM: VMBinding> GCController<VM> {
         let mut end_of_gc = EndOfGC {
             elapsed: gc_start.elapsed(),
         };
-        self.initiate_coordinator_work(&mut end_of_gc, false);
+
+        // Note: We cannot use `initiate_coordinator_work` here.  If we increment the
+        // `pending_coordinator_packets` counter when a worker spuriously wakes up, it may try to
+        // open new buckets and result in an assertion error.  That counter was there to prevent
+        // GC workers from opening more work packets while the coordinator is scheduling more work
+        // packets for the workers.  The only two other coordinator, i.e. `ScheduleCollection` and
+        // `StopMutators`, belong to that category.  `EndOfGC` doesn't add new work packets, so it
+        // doesn't need to prevent the workers from opening buckets (as there are none to be
+        // opened.)
```

Comment:
Eventually we still want to use `initiate_coordinator_work` here, right? Can you put the link to both issues in the comments, and add a `FIXME`?

Reply:
Probably not.  The `initiate_coordinator_work` function is intended for preventing workers from opening more packets.  Maybe `run_coordinator_work_and_prevent_opening_buckets` is a more meaningful (but verbose) function name.  From the diagram I drew in https://github.com/mmtk/mmtk-core/issues/774, `ScheduleCollection` and `StopMutators` are preconditions of some buckets, but `EndOfGC` is not a precondition of anything.  So it may not make sense to wrap all CoordinatorWork with the function currently named `initiate_coordinator_work`.

Anyway, I'll add links to both issues so that we can fix it more properly later.